cancat/uds.py

```diff
@@ -8,39 +8,61 @@ def __init__(self, c, tx_arbid, rx_arbid):
         self.c = c
         self.tx_arbid = tx_arbid
         self.rx_arbid = rx_arbid
-        self.t = threading.Thread(target = self.SendTesterPresent)
-        self.TesterPresent = False
-        self.t.setDaemon(True)
-        self.t.start()
 
-    def ReassembleIsoTP(self, start_index = 0):
+    def ReassembleIsoTP(self, start_index = 0, service = None):
         msg_found = False
         while msg_found is False:
             for idx, ts, arbid, msg in self.c.genCanMsgs(start=start_index, arbids=[self.rx_arbid]):
-                msg_found = True
-                return hex(arbid), msg.encode('hex')
+                # Check that the message is for the expected service, if specified
+                if service is not None:
+                    # Check if this is the right service, or there was an error
+                    if ord(msg[1]) == service or ord(msg[1]) == 0x7f:
+                        msg_found = True
+                        return arbid, msg
+                else:
+                    msg_found = True
+                    return arbid, msg
             time.sleep(0.1)
 
     def SendTesterPresent(self):
-        while True:
-            if self.TesterPresent:
-                self.c.CANxmit(self.tx_arbid, "023E000000000000".decode('hex'))
+        while self.TesterPresent is True:
+            self.c.CANxmit(self.tx_arbid, "023E000000000000".decode('hex'))
             time.sleep(2.0)
 
     def StartTesterPresent(self):
         self.TesterPresent = True
+        self.t = threading.Thread(target = self.SendTesterPresent)
```

Comment:
i understand where you're coming from.

a couple (not completely terrible, but not necessarily far off) thoughts:
* it's completely legitimate to create a "run()" function that is not part of the class def, but the run function still takes in "self".  in addition to specifying "target=run" also specify "args=[self]" or something similar.  this is not something i've tried, so it may not work.
* the atexit() function allows for cleanup when the runtime environment is about to close.  this is the place to manually "destruct" things.  this is something i've done in the past.
* weakref is a python module to handle just such things.  weakref has helpers which use weak references (ie. they point to an object but are not enough to keep it from being garbage-collected).  perhaps there's a solution in there.  perhaps use something like:
  self.t = threading.Thread(target=weakref.Proxy(self.run))

what problem are you trying to solve here?  is it the error message when closing python?
i prefer to create fewer things that hang around, and threads are heavy resources.  i prefer using one versus many when one will do the trick.  but let's talk about the problem you're solving.